### ISSUE TITLE
Macros don't correctly record mapped Ctrl keys

### ISSUE THREAD

**Issue by idbrii** [vim/vim#11270]
### Steps to reproduce

Mapping C-L to Esc doesn't work correctly in macros:

1. gvim --clean
1. `:inoremap <C-L> <Esc>`
2. populate buffer: `ivim<CR>devs<C-L>`
1. record a macro using C-L: `ggqqIhi <C-L>`
1. replay macro: `j@q`

Result:
```
hi vim
hi üLdevs
```

### Expected behaviour

When playing a macro, mappings should work the same as they while recording the macro. Using Ctrl in a mapping shouldn't replay as special characters. This was the behaviour on vim 8.2. I can verify that it works this way on 8.1.2269 (in a terminal) and I think it was also correct on 8.2.3399.

Expected:
```
hi vim
hi devs
```

### Version of Vim

9.0.636

### Environment

Win 10 21H2
gvim

Using build of vim from scoop (installer from https://github.com/vim/vim-win32-installer)

**Comment by zeertzjq** [comment]
This fixes the bug:
```diff
diff --git a/src/map.c b/src/map.c
index d063e053a..3463adcf1 100644
--- a/src/map.c
+++ b/src/map.c
@@ -1753,7 +1753,11 @@ vim_strsave_escape_csi(char_u *p)
 	d = res;
 	for (s = p; *s != NUL; )
 	{
-	    if (s[0] == K_SPECIAL && s[1] != NUL && s[2] != NUL)
+	    if ((s[0] == K_SPECIAL
+#ifdef FEAT_GUI
+		|| s[0] == CSI
+#endif
+		) && s[1] != NUL && s[2] != NUL)
 	    {
 		// Copy special key unmodified.
 		*d++ = *s++;
```

Not sure if it is possible to write a test.

**Comment by idbrii** [comment]
Wow, fast find! And I didn't realize it was gui-only, but testing in terminal vim9 I don't can't repro.

For tests, https://github.com/vim/vim/blob/master/src/testdir/test_gui.vim doesn't seem to have any macro tests. I was using `let @q = ` to convert my broken macros into working ones, so that's not a valid way to test. `feedkeys("qqIhi\<C-L>q@q")` doesn't work and `feedkeys("qqIhi\<C-L>q@q", "L")` (or "t") don't exhibit the bug.

Not sure how you can test without a script that sends keys to vim.

**Comment by zeertzjq** [comment]
Note: unlike a previous issue, the `#ifdef FEAT_GUI` in the patch above probably cannot be omitted.

With `FEAT_GUI`, `CSI` is just an ordinary character, so removing the `#ifdef FEAT_GUI` can cause problems like treating a part of a UTF-8 sequence as a part of a GUI termcap when `FEAT_GUI` is not defined.

**Comment by zeertzjq** [comment]
Note: unlike a previous issue, the `#ifdef FEAT_GUI` in the patch above probably cannot be omitted.

Without `FEAT_GUI`, `CSI` is just an ordinary character, so removing the `#ifdef FEAT_GUI` can cause problems like treating a part of a UTF-8 sequence as a part of a GUI termcap when `FEAT_GUI` is not defined.

**Comment by zeertzjq** [comment]
I guess the easiest way to test this is to just hardcode the register content. I've create a PR #11275.

**Comment by idbrii** [comment]
Fix works for me on 9.0.662. It even correctly plays macros recorded on 9.0.636.

Thanks @zeertzjq!
